### PR TITLE
[TargetParser] Define AEK_FCMA and AEK_JSCVT for tsv110

### PR DESCRIPTION
**Pull request by Qi-Hu**
We define AEK_JSCVT and AEK_FCMA for CPU features FEAT_JSCVT and FEAT_FCMA respectively, and add them to the CpuInfo of tsv110.

### CODE REVIEW

**Review by bryanpkc — APPROVED** [review]
This looks reasonable to me, but I'll wait for other reviewers to look at this. Please rebase on main and make sure the CI workflows succeed.

File: clang/lib/Driver/ToolChains/Arch/AArch64.cpp [hunk 1]
```diff
@@ -81,6 +81,15 @@ static bool DecodeAArch64Features(const Driver &D, StringRef text,
     else
       return false;
 
+    // +jsconv and +complxnum implies +neon and +fp-armv8
```
Comment (davemgreen) [inline clang/lib/Driver/ToolChains/Arch/AArch64.cpp:84]: I believe this ideally would not be in the driver, as it does not apply to target attributes, only -march options.

Comment (ilinpv) [inline clang/lib/Driver/ToolChains/Arch/AArch64.cpp:84]: I don't think we need these implications. Features dependencies are already set in llvm/lib/Target/AArch64/AArch64.td

`def FeatureComplxNum : SubtargetFeature<
    "complxnum", "HasComplxNum", "true",
    "Enable v8.3-A Floating-point complex number support (FEAT_FCMA)",
    [FeatureNEON]>;`
   
`def FeatureJS : SubtargetFeature<
    "jsconv", "HasJS", "true",
    "Enable v8.3-A JavaScript FP conversion instructions (FEAT_JSCVT)",
    [FeatureFPARMv8]>;`

Comment (Qi-Hu) [reply clang/lib/Driver/ToolChains/Arch/AArch64.cpp:84]: The dependencies between these new target attributes and others are also specified in llvm/include/llvm/TargetParser/AArch64TargetParser.h. And we want to support having nofp and nosimd disable jsconv and complxnum. This piece of code makes the driver pass the correct -target-feature flags to the backend.

Comment (momchil-velikov) [reply clang/lib/Driver/ToolChains/Arch/AArch64.cpp:84]: According to the latest Arm ARM (https://developer.arm.com/documentation/ddi0487/ja/?lang=en) the architecrtural extensions `FEAT_FCMA` and `FEAT_JSCVT` are mandatory in Armv8.3-a and are not optional in any architecture version.

For such features, our convention is to not expose them as command-line options.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h [hunk 2]
```diff
@@ -173,6 +173,8 @@ enum ArchExtKind : unsigned {
   AEK_SMEF8F16 =      69, // FEAT_SME_F8F16
   AEK_SMEF8F32 =      70, // FEAT_SME_F8F32
   AEK_SMEFA64 =       71, // FEAT_SME_FA64
+  AEK_JSCVT =         72, // FEAT_JSCVT
+  AEK_FCMA =          73, // FEAT_FCMA
```
Comment (hassnaaHamdi) [inline llvm/include/llvm/TargetParser/AArch64TargetParser.h:177]: Hi,
As you are updating this enum, I think you need to update the related tests also in the TargetParserTest.cpp, the test is named: `AArch64ExtensionFeatures`

File: clang/lib/Driver/ToolChains/Arch/AArch64.cpp (hunk 1 above)
Comment (bryanpkc) [reply clang/lib/Driver/ToolChains/Arch/AArch64.cpp:84]: @momchil-velikov Thank you for your answer. It is strange that the wording in the Arm ARM is so strict. It is well understood that an Armv8.*x*-A processor can implement any feature from the next *.x* generation (see [Understanding the Armv8.x extensions](https://developer.arm.com/-/media/Arm%20Developer%20Community/PDF/Learn%20the%20Architecture/Understanding%20the%20Armv8.x%20extensions.pdf)).

The rationale for this change is to enable the use of these ISA features on an Armv8.2-a processor (such as TSV110 and other similar chips), by making the command-line options available.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h (hunk 2 above)
Comment (Qi-Hu) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h:177]: Good points. The tests have been added.

File: clang/lib/Driver/ToolChains/Arch/AArch64.cpp (hunk 1 above)
Comment (Qi-Hu) [reply clang/lib/Driver/ToolChains/Arch/AArch64.cpp]: Indeed. The change in the driver has been removed.

Comment (momchil-velikov) [reply clang/lib/Driver/ToolChains/Arch/AArch64.cpp]: @bryanpkc Good point. The same is said also in Arm ARM, section A2.1.1.

There is already a description  at https://github.com/llvm/llvm-project/blob/ba4d36951f1deaf24dce526900a2d6dbdefa377b/llvm/lib/Target/AArch64/AArch64.td#L1513

Shouldn't it include both  `FeatureJS` and `FeatureComplxNum` ?

Once you have that, I would expect `+nosimd` to disable FEAT_FCMA and `+nofp` to disable both `FEAT_FCMA` and `FEAT_JSCVT`.

Comment (bryanpkc) [reply clang/lib/Driver/ToolChains/Arch/AArch64.cpp]: Ah yes, this PR is missing a modification in the .td file. @Qi-Hu Could youu please try that?

Comment (Qi-Hu) [reply clang/lib/Driver/ToolChains/Arch/AArch64.cpp]: Sure, it has been added.

Comment (Qi-Hu) [reply clang/lib/Driver/ToolChains/Arch/AArch64.cpp]: I have made some changes and included AEK_JSCVT and AEK_FCMA in Armv8.3-A. Could you please review it again? @davemgreen @momchil-velikov Thanks!

**Comment by bryanpkc** [comment]
@davemgreen @ilinpv @DavidSpickett @hassnaaHamdi Gentle ping.

**Review by davemgreen — APPROVED** [review]
https://github.com/ARM-software/acle/pull/279 was committed recently, where I think this lines up with the final version of it. I think this LGTM in that case.

**Pull request merged**